File: day_02/_javascript/deep.js
import fs from 'fs'

function parseInput() {
  return fs.readFileSync('../input.txt', 'utf8')
    .trim()
    .split('\n')
    .map(line => {
      const [command, value] = line.split(' ')
      return { command, value: +value }
    })
}

function solvePart1(instructions) {
  let horizontal = 0
  let depth = 0

  for (const { command, value } of instructions) {
    if (command === "forward") {
      horizontal += value
    } else if (command === "down") {
      depth += value
    } else if (command === "up") {
      depth -= value
    }
  }

  return horizontal * depth
}

function solvePart2(instructions) {
  let aim = 0
  let horizontal = 0
  let depth = 0

  for (const { command, value } of instructions) {
    if (command === "forward") {
      horizontal += value
      depth += aim * value
    } else if (command === "down") {
      aim += value
    } else if (command === "up") {
      aim -= value
    }
  }

  return horizontal * depth
}

const instructions = parseInput()
const part1 = solvePart1(instructions)
const part2 = solvePart2(instructions)

console.log(part1)
console.log(part2)

File: kotlin/day02/deep.kt
package day02

import java.io.File

data class Instruction(val command: String, val value: Int)

fun main() {
    val instructions = parseInput()
    val part1 = solvePart1(instructions)
    val part2 = solvePart2(instructions)
    println(part1)
    println(part2)
}

fun parseInput(): List<Instruction> {
    return File("../../input/02.txt")
        .readLines()
        .map { it.split(" ") }
        .map { Instruction(it[0], it[1].toInt()) }
}

fun solvePart1(instructions: List<Instruction>): Int {
    var horizontal = 0
    var depth = 0

    instructions.forEach {
        val (command, value) = it
        when (command) {
            "forward" -> horizontal += value
            "down" -> depth += value
            "up" -> depth -= value
        }
    }

    return horizontal * depth
}

fun solvePart2(instructions: List<Instruction>): Int {
    var aim = 0
    var horizontal = 0
    var depth = 0

    instructions.forEach {
        val (command, value) = it
        when (command) {
            "forward" -> {
                horizontal += value
                depth += aim * value
            }
            "down" -> aim += value
            "up" -> aim -= value
        }
    }

    return horizontal * depth
}
